server/src/main/java/io/crate/analyze/AlterTableRenameColumnAnalyzer.java

```diff
@@ -89,6 +90,11 @@ public AnalyzedAlterTableRenameColumn analyze(AlterTableRenameColumn<Expression>
         ColumnIdent sourceCol = sourceRef.column();
         ColumnIdent targetCol = targetRef.column();
 
+        if (sourceRef.oid() == Metadata.COLUMN_OID_UNASSIGNED) {
+            throw new IllegalStateException(
+                String.format(Locale.ENGLISH, "Cannot rename a column that does not have an OID assigned: %s", sourceCol));
+        }
+
```

Comment:
Hi @BaurzhanSakhariev do you think this logic makes sense? Are there anything else I should look into regarding oids? Thank you.

Reply:
>Are there anything else I should look into regarding oids? 

I think all `DocTableInfo.versionCreated` usages are not safe for partitioend tables
 (`Indexer.writeOids` for example).


Another usage: `TransportAddColumnAction` that uses `versionCreated` to assign/not assign OID.

Basically, we might NOT assign oid for tabes where we could have done, so, maybe such approach could work for DDL statements but let me check all usages to see if we can have a more generic fix. 

Also, if a partitioned table was ALTERED and faced a bug fixed in 5.9.7 (https://github.com/crate/crate/pull/17178), even shardCreatedVersion usage might be not enough, ie we would need https://github.com/crate/crate/pull/17186.

But let me double check all this assessments.

Reply:
>Also, if a partitioned table was ALTERED and faced a bug fixed in 5.9.7 (https://github.com/crate/crate/pull/17178), even shardCreatedVersion usage might be not enough, ie we would need https://github.com/crate/crate/pull/17186.

Not true for shard version related use cases, see https://github.com/crate/crate/pull/17203#issuecomment-2595297051.

Only table version related logic (so this PR's use case, DDL) is affected by the ALTER TABLE bug but as [mentioned](https://github.com/crate/crate/pull/17216#pullrequestreview-2552436157) below, we need a universal solution, so probably something like https://github.com/crate/crate/pull/17186

Reply:
> we need a universal solution, so probably something like https://github.com/crate/crate/pull/17186

BTW, https://github.com/crate/crate/pull/17186 won't help those partitioned tables with old partitions dropped. i.e. `altered` tableVersionCreated and only new partition versions remaining. There is no trace of older version number anywhere.

Reply:
@mfussenegger looks like then we have to add special handling to table version related logic (ADD/DROP/RENAME)? I don't see how we can fix broken version setting for existing partitions for the case mentioned above.

Reply:
We could consider using the fact that any `ref.oid() == Metadata.COLUMN_OID_UNASSIGNED` implies that `tableVersionCreated` must be < `5.5` when implementing the metadata upgrader.

Reply:
I also looked into all usages of `docTableInfo.versionCreated()`
```
- addColumnAction/Indexer.writeOids (passed to DynamicIndexer only)
    - the upgraded versionCreated may cause the new columns to use oids but I dont see any functional failures that could be caused by these
- drop/renameActions
    - they will cause metadata corruptions
- ColumnFieldVisitor
    - this is a new(unreleased) feature. The versionCreated is used as `in.setVersion(tableVersion)`.
```
Looking at them, I think only extra checks needed are rename and drop.
I sort of see why fixing the version settings in templateMetadata could be unsafe. In order to make sure that it is safe, we need to check all usages of `docTableInfo.versionCreated()` for all released CrateDB versions.